Documentation for stubbing modules

kobbikobb opened the pull request:
It is not obvious from the documentation how you stub external dependencies (modules that are imported from the modules under test).
I feel we should add something to help users with this common use case.
What do you think?

 #### Purpose (TL;DR) - mandatory
<!--
> give a concise (one or two short sentences) description of what what problem is being solved by this PR
>
> Example: Fix issue #123456 by re-structuring the colour selection conditional in method `paintBlue`
-->


<!--
 #### Background (Problem in detail)  - optional
-->
<!--
> When relevant, give a more thorough description of what the problem the PR is trying to solve. Examples of good topics for this section are:
> * Link to an existing GitHub issue describing the problem
> * Describing the problem in greater detail than the TL;DR section above
> * How you discovered the issue, if it's not already described as an issue on GitHub
> * Discussion of different approaches to solving this problem and why you chose your proposed solution
-->


<!--
 #### Solution  - optional
-->
<!--
> When contributing code (and not just fixing typos, documentation and configuration), please describe why/how your solution works. This helps reviewers spot any mistakes in the implementation.
>
> Example:
> "This solution works by adding a `paintBlue()` method"
> Then your reviewer might spot a mistake in the implementation, if `paintBlue()` uses the colour red.
-->

 #### How to verify - mandatory
1. Check out this branch
2. `npm install`
3. <your-steps-here>

 #### Checklist for author

- [x] `npm run lint` passes
- [x] References to standard library functions are [cached](https://github.com/sinonjs/sinon/pull/1523).

fatso83 commented:
I think this should go here: https://github.com/sinonjs/sinon/tree/master/docs/_howto

With a link from the stub docs.

kobbikobb commented:
@fatso83 good idea, I've pushed changes :)

fatso83 commented on docs/_howto/stub-dependency.md:
```diff
@@ -0,0 +1,21 @@
+title: How to stub a dependency of a module?
+---
+
+To stub dependency(imported module) of a module under test you have to import it explicitly in your test and stub the desired method.
```
missing a space.

fatso83 replied:
You might add that Sinon is simply a stubbing library, not a module intercepting library, and how you stub out dependencies is highly dependant on your environment. For Node environments, we usually recommend solutions targetting [link seams](https://github.com/sinonjs/sinon/blob/8683af388a2b46b2b3e7764fe6fcfe91b8543e35/docs/_howto/link-seams-commonjs.md) (replace with relative link) or explicit dependency injection. Though in _some simple cases_, you can get away with just using Sinon by modifying the module exports of the dependency. 

(_And then comes your text_)

fatso83 commented on docs/_howto/stub-dependency.md:
```diff
@@ -0,0 +1,21 @@
+import * as importedModule from "./importedModule";
+import {getTheSecret} from "./moduleUnderTest";
+
+"test should get secret": function () {
```
This isn't actually valid javascript as it is and will result in a runtime error (`Unexpected token, expected ;`). Could you change it to the standard BDD style Mocha test style we use everywhere else?

fatso83 commented on docs/_releases/latest/stubs.md:
```diff
@@ -76,6 +76,8 @@ Replaces `object.method` with a stub function. An exception is thrown if the pro
 
 The original function can be restored by calling `object.method.restore();` (or `stub.restore();`).
 
+A dependency of a module can be stubbed [see example](../../_howto/stub-dependency.md).
```
```suggestion
Stubs can be used to replace a module's dependencies, ensuring it is tested in isolation. [For simple cases](../../_howto/stub-dependency.md) you might not need additional helper libraries, but for more complex cases you might need to [target the module loading mechanisms](../../link-seams-commonjs.md).
```

kobbikobb commented:
@fatso83 thank you for your comments and suggestions. I have done your suggestions and added also the modules under test. Mocking modules this way is highly dependant upon the implementation, so adding a complete example might help the user understand better.

Any counter proposals?
I also rebased the code...

fatso83 commented on docs/_howto/stub-dependency.md:
```diff
@@ -0,0 +1,54 @@
+
+Sinon is simply a stubbing library, not a module interception library. Stubbing dependencies is highly dependant on your enviroment and the implementation. For Node environments, we usually recommend solutions targetting [link seams](./link-seams-commonjs) or explicit dependency injection. Though in some simple cases, you can get away with just using Sinon by modifying the module exports of the d ...
+
+To stub a dependency (imported module) of a module under test you have to import it explicitly in your test and stub the desired method. For the mocking to work, the mocked method cannot be deconstructed, neither in the module under test nor in the test.
```
I'd replace "mocking" with "stubbing" and "mocked" with "stubbed", since we actually differ "mocks" from "stubs" in our docs and API. Otherwise, people might get a bit confused.

Also, replace "deconstructed" with "destructured", using a link to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment). That's assuming I understand it's actually destructuring you refer to? As in
```javascript
const {a} = require('/module-with-a-export');
```

kobbikobb replied:
@fatso83 Makes sense, yes you are right about destructed, and the link to MDN makes this more clear. Good feedback :)

I've updated the code accordingly.

BTW, I see there have already been discussions about this topic.
https://github.com/sinonjs/sinon/issues/1538

Hopefully this documentation we are adding will help developers having problems with stubbing dependencies this way :)

fatso83 commented:
I'm assuming this is good enough for merge, since no-one has chimed in. It adds value and if anyone thinks it can be improved, we now have a good foundation to be build on for further additions :-)

Thanks for the work and sticking it through!
